neonvm/runner/containermgr.go

```diff
@@ -0,0 +1,364 @@
+	}
+
+	var httpPort int
+	flag.IntVar(&httpPort, "port", -1, "Port for the CPU http server")
+	flag.Parse()
+
+	var pods crictlPods
+	if err := crictl.run(logger, &pods, "pods", "-o", "json"); err != nil {
+		logger.Fatal("failed to run crictl command to get CRI ID for pod UID", zap.String("uid", selfPodUID), zap.Error(err))
+	}
+
+	// find pod with matching uid
+	var criPodID string
+	for _, p := range pods.Items {
+		if p.Metadata.UID == selfPodUID {
+			criPodID = p.ID
+			break
+		}
+	}
+	if criPodID == "" {
+		logger.Fatal("could not find CRI pod with matching UID", zap.String("uid", selfPodUID))
+	}
+
+	logger.Info("Got CRI ID for pod", zap.String("podID", criPodID))
+
+	var criRunnerContainerID string
+	for criRunnerContainerID == "" {
+		var containers crictlContainers
+		if err := crictl.run(logger, &containers, "ps", "-p", criPodID, "-o", "json"); err != nil {
+			logger.Fatal(
+				"failed to run crictl command to get CRI ID for container",
+				zap.String("podID", criPodID),
+				zap.String("name", runnerContainerName),
+				zap.Error(err),
+			)
+		}
+
+		for _, c := range containers.Containers {
+			if c.Metadata.Name == runnerContainerName {
+				criRunnerContainerID = c.ID
+				break
+			}
+		}
+		if criRunnerContainerID == "" {
+			logger.Error(
+				"could not find CRI container with matching name",
+				zap.String("podID", criPodID),
+				zap.String("name", runnerContainerName),
+			)
+			time.Sleep(retryInitEvery)
+		}
+	}
+
+	logger.Info(
+		fmt.Sprintf("Got CRI ID for %s container", runnerContainerName),
+		zap.String("containerID", criRunnerContainerID),
+	)
+
+	cinfo := containerInfo{
+		podID:        criPodID,
+		containerID:  criRunnerContainerID,
+		lastMilliCPU: atomic.Uint32{},
+	}
+
+	cinfo.listenForCPUChanges(context.TODO(), logger, crictl, int32(httpPort))
+}
+
+type crictlConfig struct {
+	endpoint string
+}
+
+func (c *crictlConfig) run(logger *zap.Logger, output any, args ...string) error {
+	actualArgs := []string{
+		"--runtime-endpoint",
+		c.endpoint,
+	}
+	actualArgs = append(actualArgs, args...)
+
+	logger.Info("running crictl", zap.Strings("args", actualArgs))
+
+	cmd := exec.Command("/usr/bin/crictl", actualArgs...)
+
+	stderr, err := os.OpenFile("/dev/stderr", os.O_RDWR, 0 /* unused */)
+	if err != nil {
+		panic(fmt.Errorf("failed to open /dev/stderr: %w", err))
+	}
+	cmd.Stderr = stderr
+
+	if output == nil {
+		if err := cmd.Run(); err != nil {
+			return fmt.Errorf("failed to run command: %w", err)
+		}
+		return nil
+	} else {
+		out, err := cmd.Output()
+		if err != nil {
+			return fmt.Errorf("failed to run command: %w", err)
+		}
+
+		if err := json.Unmarshal(out, output); err != nil {
+			return fmt.Errorf("error parsing JSON: %w", err)
+		}
+		return nil
+	}
+}
+
+type containerInfo struct {
+	podID        string
+	containerID  string
+	lastMilliCPU atomic.Uint32
+}
+
+func (c *containerInfo) listenForCPUChanges(ctx context.Context, logger *zap.Logger, crictl *crictlConfig, port int32) {
```

Comment:
This name is also a bit confusing. We provide an interface to get and set CPU limits rather listening on cpu changes. Moreover, here we handle health check requets as well that is not related.

Reply:
Copied from neonvm-runner, happy to address in a follow-up.

https://github.com/neondatabase/autoscaling/blob/32e407a6ba1ae49c341e9de898d62489914fff74/neonvm/runner/main.go#L774